Allow XrootD for stageout

Rumor has it that there is some way to use XrootD to send files in, not just copy them out. We should add XrootD to our stageout handling if so.

On slack, @matz-e demonstrated that this seems to be a configuration issue at ND:
```
earth: xrdcp rootlogon.C root://eoscms.cern.ch//eos/cms/store/user/matze/foo3?svcClass=default
[3.148kB/3.148kB][100%][==================================================][3.148kB/s[3.148kB/3.148kB][100%][==================================================][1.574kB/s]
earth:

matze [11:00 AM]
lxplus0017: eos
EOS Console [root://eoscms.cern.ch] |/> ls /store/user/matze
TTbarH_M-125_13TeV_amcatnlo-pythia8-tauola
foo
foo3
test_se
ttH
EOS Console [root
```

And that permission management seems to be active at CERN:
```
earth: xrdcp rootlogon.C root://eoscms.cern.ch//eos/cms/store/user/muell149/greetings_charlie?svcClass=default
[0B/0B][100%][==================================================][0B/s]
Run: [ERROR] Server responded with an error: [3010] Unable to open file /eos/cms/store/user/muell149/greetings_charlie; Operation not permitted
```

To be fair, all that's been demonstrated so far is that it's support at 
CERN and not ND.  We don't know yet whether ND or CERN is the more 
common case at the moment.  I guess there's rumors this is support at 
FNAL too?

On 7/29/15 11:15 AM, Anna Woodard wrote:
>
> On slack, @matz-e <https://github.com/matz-e> demonstrated that this 
> seems to be a configuration issue at ND:
>
> |earth: xrdcp rootlogon.C 
> root://eoscms.cern.ch//eos/cms/store/user/matze/foo3?svcClass=default 
> [3.148kB/3.148kB][100%][==================================================][3.148kB/s[3.148kB/3.148kB][100%][==================================================][1.574kB/s] 
> earth: matze [11:00 AM] lxplus0017: eos EOS Console 
> [root://eoscms.cern.ch] |/> ls /store/user/matze 
> TTbarH_M-125_13TeV_amcatnlo-pythia8-tauola foo foo3 test_se ttH EOS 
> Console [root |
>
> And that permission management seems to be active at CERN:
>
> |earth: xrdcp rootlogon.C 
> root://eoscms.cern.ch//eos/cms/store/user/muell149/greetings_charlie?svcClass=default 
> [0B/0B][100%][==================================================][0B/s] Run: 
> [ERROR] Server responded with an error: [3010] Unable to open file 
> /eos/cms/store/user/muell149/greetings_charlie; Operation not permitted |
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/matz-e/lobster/issues/161#issuecomment-125984835>.
>

--- a/lobster/core/data/task.py
+++ b/lobster/core/data/task.py
@@ -739,6 +739,27 @@ def copy_outputs(data, config, env):
                     break
                 else:
                     data['transfers'][protocol]['failure'] += 1
+            elif output.startswith("root://"):
+                args = [
+                    "env",
+                    "XRD_LOGLEVEL=Debug",
+                    "xrdcp",
+                    localname,
+                    os.path.join(output, remotename)
+                ]
+                p = run_subprocess(args)
+                logger.info('Checking output file transfer.')
+                if p.returncode == 0 and check_output(config, localname, remotename):
+                    logger.info('File transfer successful!')
+                    transferred.append(localname)
+                    match = server_re.match(args[-1])
+                    if match:
+                        target_se.append(match.group(1))
+                    data['transfers']['root']['stageout success'] += 1
+                    break
+                else:
+                    data['transfers']['root']['stageout failure'] += 1
+               
             elif output.startswith("chirp://"):
                 server, path = re.match("chirp://([a-zA-Z0-9:.\-]+)/(.*)", output).groups()
 
@@ -764,6 +785,7 @@ def copy_outputs(data, config, env):
                     break
                 else:
                     data['transfers']['chirp']['stageout failure'] += 1
+
             elif output.startswith("hdfs://"):
                 server, path = re.match("hdfs://([a-zA-Z0-9:.\-]+)/(.*)", output).groups()
                 server = "hdfs://" + server

